### PR TITLE
Port WaveNet to PyTorch

### PR DESCRIPTION
**Pull request by abdulfatir**
*Description of changes:*
This PR adds a PyTorch version of the WaveNet model.

Comparison between MxNet and PyTorch versions on the **solar** dataset:

| model         |   NRMSE |     ND |   mean_wQuantileLoss |   sMAPE |   MASE |
|:--------------|--------:|-------:|---------------------:|--------:|-------:|
| wavenet-mxnet |  0.9608 | 0.4897 |               0.3621 |  1.3701 | 1.1474 |
| wavenet-torch |  0.9517 | 0.4825 |               0.3679 |  1.3669 | 1.1296 |

Each model was trained for 10 epochs. Scores have been averaged over 3 runs.

TODOs:
- [x] Initial model port
- [x] Comparison with MxNet version 
- [x] Add tests
- [x] Add doctrings and type hints
- [x] Test on GPU
- [x] Update available models

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by abdulfatir** [comment]
**UPDATE**: The test passes upon summing over the last dimension. Thanks @shchur!

I am still confused and unsure about the causality aspect of this model. Networks have been ported from the MXNet version (I may have made a mistake during the port). 

I ran the following test to check whether the model is "causal": If a model is causal, then gradients of the output at a specific `time_step` w.r.t. the input should be zero for all timesteps after `time_step`. It turns out that this is not true for this implementation. For comparison, I also include similar tests for a Causal Conv stack and an RNN.

```py
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F

from gluonts.torch.model.wavenet import WaveNetEstimator

torch.manual_seed(42)
np.random.seed(42)


class CausalConv1d(nn.Module):
    def __init__(
        self,
        in_channels: int,
        out_channels: int,
        kernel_size: int,
        dilation: int,
    ):
        super().__init__()

        self.conv = nn.Conv1d(
            in_channels,
            out_channels,
            kernel_size=kernel_size,
            dilation=dilation,
        )
        self.padding = (kernel_size - 1) * dilation

    def forward(self, sequences: torch.Tensor) -> torch.Tensor:
        padded_sequences = F.pad(sequences, pad=(self.padding, 0))
        out = self.conv(padded_sequences)
        return out


# Setup WaveNet
est = WaveNetEstimator("H", prediction_length=64)
wavenet = est.create_lightning_module().model

# Setup CausalConv stack
dilations = [1, 2, 4, 8, 16, 32, 64, 128, 256]
causal_conv = nn.Sequential(*[CausalConv1d(24, 24, 2, d) for d in dilations])

# Setup RNN
rnn = nn.GRU(input_size=24, hidden_size=24, num_layers=4, batch_first=True)


# (batch, features, time)
x = torch.rand(1, 24, est.context_length + est.prediction_length - 1)
x.requires_grad_(True)
time_step = -10

# Feed into WaveNet and sum over feature dimensions
y = wavenet.base_net(x)[0].sum(-2)  # (batch, time)
# Grad of output at time_step w.r.t. to input summed over feature dimension
grad = torch.autograd.grad(y[0, time_step], x)[0].sum(-2)
# For the model to have "causal" property, grad of all timesteps after time_step should be zero
grad_is_zero = all(grad[0, time_step + 1 :] == 0)
print("WaveNet:", grad_is_zero)

# Feed into CausalConv stack and sum over feature dimensions
y = causal_conv(x).sum(-2)  # (batch, time)
# Grad of output at time_step w.r.t. to input summed over feature dimension
grad = torch.autograd.grad(y[0, time_step], x)[0].sum(-2)
grad_is_zero = all(grad[0, time_step + 1 :] == 0)
print("CausalConv stack:", grad_is_zero)

# Feed into RNN and sum over feature dimensions
y = rnn(x.transpose(1, 2))[0].sum(-1)  # (batch, time)
# Grad of output at time_step w.r.t. to input summed over feature dimension
grad = torch.autograd.grad(y[0, time_step], x)[0].sum(-2)
grad_is_zero = all(grad[0, time_step + 1 :] == 0)
print("RNN:", grad_is_zero)
```
**Output**:
```
WaveNet: False
CausalConv stack: True
RNN: True
```

cc @shchur @vafl @rshyamsundar @kashif @lostella @shubhamkapoor

**Comment by abdulfatir** [comment]
Ran the same test for MXNet 🫠:

```py
import mxnet as mx
from gluonts.mx import WaveNetEstimator

est = WaveNetEstimator("H", prediction_length=64)
wavenet = est.create_training_network()
wavenet.initialize()

x = mx.nd.random.uniform(shape=(1, 24, 535))
x.attach_grad()
time_step = -10

with mx.autograd.record():
    y = wavenet.base_net(mx.nd, x)[0].sum(-2)[0, time_step]
dx = mx.autograd.grad(y, [x], create_graph=True)[0].sum(-2)

grad_is_zero = all(dx[0, time_step + 1 :] == 0)
print("WaveNet:", grad_is_zero)

rnn = mx.gluon.rnn.GRU(
    hidden_size=24, input_size=24, num_layers=4, layout="NTC"
)
rnn.initialize()
with mx.autograd.record():
    y = rnn(x.transpose((0, 2, 1))).sum(-1)[0, time_step]
dx = mx.autograd.grad(y, [x], create_graph=True)[0].sum(-2)

grad_is_zero = all(dx[0, time_step + 1 :] == 0)
print("RNN:", grad_is_zero)
```

**Output**:
```
WaveNet: False
RNN: True
```

File: src/gluonts/torch/model/wavenet/module.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+            else:
+                trimmed_skip = skip[..., self.trim_lengths[i] :]
+            skip_outs.append(trimmed_skip)
+
+        y = torch.stack(skip_outs).sum(dim=0)
+        y = self.output_act(y)
+        y = self.conv1(y)
+        y = self.output_act(y)
+        y = self.conv2(y)
+        logits = y.transpose(1, 2)
```
Comment (shchur) [inline src/gluonts/torch/model/wavenet/module.py:373]: @abdulfatir I think the problem in your test appears because of this transpose here: if you compute the grad of `y` w.r.t. `inputs`, you will get correct causal masking.

Comment (abdulfatir) [reply src/gluonts/torch/model/wavenet/module.py:373]: Ah, right! I need to sum over `-1`, not `-2`.

Comment (vafl) [reply src/gluonts/torch/model/wavenet/module.py:373]: I really like this test @abdulfatir (and good catch @shchur). I remember we had off-by-one errors a few times between training and prediction target. Maybe one can make the test re-usable for other models and also check that no time point is skipped.

**Review by vafl — APPROVED** [review]
Looks good to me. Thanks @abdulfatir

**Comment by abdulfatir** [comment]
Thank you for reviewing, @vafl!

File: src/gluonts/torch/model/wavenet/module.py (hunk 1 above)
Comment (abdulfatir) [reply src/gluonts/torch/model/wavenet/module.py:373]: That's a good idea. Let me check how we can add these tests to other models.

Comment (lostella) [reply src/gluonts/torch/model/wavenet/module.py:373]: > I remember we had off-by-one errors a few times

Two hard things in computer science: cache invalidation, naming things, and off-by-one errors
